### DISCUSSION TITLE
Example for commenting on a post

### DISCUSSION THREAD

**Discussion by yusupm** [sbertix/Swiftagram#159] [Q&A] [answered]
Hi sbertix, hope you have had a great weekend.

I was wondering if you are ever planning to make videos that explaining Swiftagram. I am very new to Swift and am having difficulties understanding how Swiftagram works so sorry for a newbie question. 

My main question was if you could show me and example where you use the secret to comment on a post.

Thanks in advance.

**Comment by sbertix** [comment, marked as answer]
Hey @yusupm, 

I did indeed have a great weekend, thx man 😊 Hope you had one too. 

> I was wondering if you are ever planning to make videos that explaining Swiftagram.

I never actually thought about it. I mean it would be nice having like introductory video-tutorials, but I'm honestly not sure I have the time to actually go through with it. 
It's more likely I'll just write more examples and explanations (it's already planned for `5.0.0`). 

> My main question was if you could show me and example where you use the secret to comment on a post.

Sure. I'm gonna assume you're using **Combine**, and that you don't already know the `identifier` of the media you wanna comment on. 
I'll just assume, for simplicity, you wanna comment "Amazing picture" on Instagram's last post, without you even knowing their user `identifier`, in order to be as in depth as I can be. 

```swift
import Combine
import Foundation

import Swiftagram

/// A `class` we're defining just to hold reference to all calls. 
/// This is not necessary is just to make sure you understand
/// `bin` has to be retained. 
final class Instagram {
   /// A shared instance of `Instagram`.
   static let `default`: Instagram = .init()
   /// The dispose bag.
   private var bin: Set<AnyCancellable> = []

   /// Init. 
   private init() { }

   /// Comment "Amazing picture" on Instagram's last post. 
   ///
   /// - parameter secret: A valid `Secret`.
   func comment(with secret: Secret) {
        // Return all users matching the query
        // `Instagram`.
        Endpoint.User.all(matching: "Instagram")
            .unlocking(with: secret)
            .publish()
            .prefix(1)  // Fetch one page only.
            .compactMap(\.users?.first?.identifier) // @instagram has to be the first one.
            .flatMap { instagramUserIdentifier in
                // Request one page of the most recent
                // posts made by @instagram.
                Endpoint.Media.Posts.owned(by: instagramUserIdentifier)
                    .unlocking(with: secret)
                    .publish()
                    .prefix(1)  // One page is more than enough as we only need one post.
                    .compactMap(\.media?.first?.identifier) // Get the identifier for the most recent one.
            }
            .flatMap { mostRecentInstagramPostIdentifier in
                // Post a comment.
                Endpoint.Media.Posts.comment("Amazing picture", on: mostRecentInstagramPostIdentifier)
                    .unlocking(with: secret)
                    .publish()
            }
            .sink(receiveCompletion: { if case .failure(let error) = $0 { print(error.localizedDescription) }},
                  receiveValue: { if let error = $0.error { print(error.localizedDescription) }})
            .store(in: &bin)
   }

   /// Comment "Amazing picture" on Instagram's last post. 
   ///
   /// - parameter secret: An optional `Secret`.
   func comment(with secret: Secret?) {
        guard let secret = secret else { return }
        comment(with: secret)
   }
}
```

This way you can do `Instagram.default.comment(with: secret)` anywhere in your code, to post that comment. 
Of course you can tweak the implementation so you can pass your own comment, media identifier or even username reference, depending on what data you have available. 
And again the whole `Instagram` wrapper and the `.default` thing is just so it's clear `bin` has to be retained. You should not actually rely on it in your code (or at least there are better ways to do it in an actual project). 
I hope it's clear enough.

**Reply by yusupm** [reply]
> I did indeed have a great weekend, thx man 😊 Hope you had one too.

I have also had an amazing weekend, thanks for asking


>This way you can do Instagram.default.comment(with: secret) anywhere in your code, to post that comment.

Thanks for the great example, this has covered most of what I needed. I am a bit confused on these two lines `let ``default``: Instagram = .init()` and `private var bin: Set<AnyCancellable> = []` could not understand the need and use of these variables. Would you mind explaining them?

Also is KeychainStorage something like a database that stores encrypted version of user credentials? Is it possible to call the secret inside KechainStorage for a specific account with username (example `KeychainStorage.find(with: username")`).

Sorry for asking too many questions :)

**Reply by sbertix** [reply]
> Thanks for the great example, this has covered most of what I needed. I am a bit confused on these two lines let ``default``: Instagram = .init() and private var bin: Set<AnyCancellable> = [] could not understand the need and use of these variables. Would you mind explaining them?

The `default` is just to create a shared instance of `Instagram` that is retained throughout the execution of your process. 
`Instagram(secret: secret).comment()`, for instance, wouldn't work as the `Instagram` instance is released when you reach the end of your current scope (e.g. what's "inside" of a function if you're calling that from within a function). 
The reason why it has to be retained is exactly cause `Set<AnyCancellable>` must be. Using `Combine`, if you're not retaining your cancellable set, your publisher stream will be cancelled immediately on reaching the end of the current scope, i.e. the entire `comment` logic wouldn't be executed if it wasn't for the retained `bin`. 

> Also is KeychainStorage something like a database that stores encrypted version of user credentials? Is it possible to call the secret inside KechainStorage for a specific account with username (example KeychainStorage.find(with: username")).

`KeychainStorage` is indeed a way to deal with the storing of `Secret`s automatically. However, as the username might change and it's unknown at the time of login, you can only find `Secret`s by their user identifier. So `KeychainStorage.find(with: "25025320")`, in case you're logged in with the `@instagram` account (as `25025320` is their user identifier).

**Reply by yusupm** [reply]
Oh it makes sense now. 

I have managed to make it work at the end even though I had to change few parts because of errors. Do you have any idea why I am facing the following errors:  `Instance member 'default' cannot be used on type 'Instagram'; did you mean to use a value of this type instead?` when trying `Instagram.default.comment(with: secret)` on my code after inserting the code you have given me into a swift file and using it on my code in this way `Instagram.default.comment(with: secret).` Another error is  `Value of optional type 'Secret?' must be unwrapped to a value of type 'Secret'` get this when I do `.unlocking(with: secret)`  and the error goes away when I insert `!` after `secret` .

As I said I have managed to make the code work at the end with small tweaks so these errors are not a big deal just asking to get a better understanding.

> `KeychainStorage` is indeed a way to deal with the storing of `Secret`s

Thanks for clarifying.

Also is it possible to get the link of the profile picture of a user who just logged in?

**Reply by sbertix** [reply]
> I am facing the following errors:  `Instance member 'default' cannot be used on type 'Instagram'; did you mean to use a value of this type instead?` 

Yep, I forgot the `static` keyword in the definition's last iterations. Fixed. 

> Another error is  `Value of optional type 'Secret?' must be unwrapped to a value of type 'Secret'` get this when I do `.unlocking(with: secret)`  

You're simply trying to pass an `Optional<Secret>` in place of a `Secret`. 
Please refer to the Swift docs and guides on the use of optionals as it's definitely a prerequisite. 

> and the error goes away when I insert `!` after `secret` .

You should never force unwrap optionals. 
I've added a function overload for optional `Secret`s, unwrapping the argument "the right way". 
So now you can update your code, remove the `!` and it will still work. 

Let me know if it's all clear 😊

**Reply by yusupm** [reply]
I just had a chance to look at your code and it works flawless now, thanks.  I also wanted to ask you about ways of storing user identifier once user logs in using BasicAuthenticator. I have tried to save it in an array to store multiple account's identifier by simply appending `KeychainStorage.all().last.identifier` amongst other details like username. It has worked for a while, but sometimes a user in a list has the same identifier as the other user that is already logged in. This happens even though I add to the list once the authentication is complete with success.

Would be happy if you could share some tips on proper ways of storing the identifier of users to avoid logging in.

By the way thanks for dealing with me for this long , this wrapper is just too good I want to know everything😅.

**Reply by sbertix** [reply]
> I just had a chance to look at your code and it works flawless now, thanks. I also wanted to ask you about ways of storing user identifier once user logs in using BasicAuthenticator. I have tried to save it in an array to store multiple account's identifier by simply appending KeychainStorage.all().last.identifier amongst other details like username. It has worked for a while, but sometimes a user in a list has the same identifier as the other user that is already logged in. This happens even though I add to the list once the authentication is complete with success.

Thing is `KeychainStorage.all()` is not guaranteed to return them in any given order, so you shouldn't rely on the last one, actually being the last one added. 
Instead, what you should do, is deal with the completion handler in `BasicAuthenticator` (or any other `Authenticator`) to add the newly authenticated `Secret` to your "identifier cache", for instance `UserDefaults`. 

```swift
BasicAuthenticator(storage: KeychainStorage<Secret>(), username: "", password: "")
   .authenticate {
      // We only care about secrets for now so let's assume you already
      // have the code to deal with challenges, 2fa and errors.
      guard let secret = try? $0.get() else { return }
      // Obtain the current logged in array and add the new identifier. 
      let identifiers = (UserDefaults.standard.object(forKey: "loggedInUsers") as? [String] ?? [])+[secret.identifier]
      // Update the cached version.
      UserDefaults.standard.set(identifiers, forKey: "loggedInUsers")
      UserDefaults.standard.synchronize()
   }
```

Then when you have to retrieve, for instance, the last authenticated `Secret`, all you have to do is… 

```swift
guard let lastIdentifier = (UserDefaults.standard.object(forKey: "loggedInUsers") as? [String])?.last,
    let secret = KeychainStorage<Secret>().find(with: lastIdentifier) else { return }
// Now you have your `Secret` you can pass to any request.
```

> By the way thanks for dealing with me for this long , this wrapper is just too good I want to know everything😅.

Hahaha no worries. Thank you man, you're too kind :blush:

**Reply by yusupm** [reply]
You are a legend 👑, everything is working as I wanted . 

I am very excited for 5.0 release, can't wait to see new features.

Thank you for your help and wishing you good luck.
